pallets/connectors/src/message.rs

```diff
@@ -0,0 +1,277 @@
+use common_traits::Moment;
+use frame_support::BoundedVec;
+use sp_std::vec;
+use sp_std::vec::Vec;
+
+use crate::*;
+
+#[derive(Decode, Clone, PartialEq, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub enum Message<Domain, PoolId, TrancheId, Balance, Rate>
+where
+	Domain: Encode + Decode,
+	PoolId: Encode + Decode,
+	TrancheId: Encode + Decode,
+	Balance: Encode + Decode,
+	Rate: Encode + Decode,
+{
+	Invalid,
+	AddPool {
+		pool_id: PoolId,
+	},
+	AddTranche {
+		pool_id: PoolId,
+		tranche_id: TrancheId,
+		token_name: [u8; 32],
+		token_symbol: [u8; 32],
+	},
+	UpdateTokenPrice {
+		pool_id: PoolId,
+		tranche_id: TrancheId,
+		price: Rate,
+	},
+	UpdateMember {
+		pool_id: PoolId,
+		tranche_id: TrancheId,
+		address: [u8; 32],
+		valid_until: Moment,
+	},
+	Transfer {
+		pool_id: PoolId,
+		tranche_id: TrancheId,
+		domain: Domain,
+		destination: [u8; 32],
+		amount: Balance,
+	},
+}
+
+impl<
+		Domain: Encode + Decode,
+		PoolId: Encode + Decode,
+		TrancheId: Encode + Decode,
+		Balance: Encode + Decode,
+		Rate: Encode + Decode,
+	> Message<Domain, PoolId, TrancheId, Balance, Rate>
+{
+	fn call_type(&self) -> u8 {
+		match self {
+			Self::Invalid => 0,
+			Self::AddPool { .. } => 1,
+			Self::AddTranche { .. } => 2,
+			Self::UpdateTokenPrice { .. } => 3,
+			Self::UpdateMember { .. } => 4,
+			Self::Transfer { .. } => 5,
+		}
+	}
+}
+
+impl<
+		Domain: Encode + Decode,
+		PoolId: Encode + Decode,
+		TrancheId: Encode + Decode,
+		Balance: Encode + Decode,
+		Rate: Encode + Decode,
+	> Encode for Message<Domain, PoolId, TrancheId, Balance, Rate>
+{
+	fn encode(&self) -> Vec<u8> {
```

Comment:
I think we should only implement `using_encode` as the doc mention. Assuming, this is a value type, and I understand the definition of value vs. allocating type.

> /// Implementations should override `using_encoded` for value types and `encode_to` and `size_hint` for allocating types.
> /// Wrapper types should override all methods.

Reply:
@mustermeiszer thank for the input. `using_encoded` has a completely different api that doesn't seem to quite fit what we need here. We need to go from `Message` to `Vec<u8>`; that's the api that `encode` provides whereas `using_encoded` is `using_encoded<R, F: FnOnce(&[u8]) -> R>(&self, f: F) -> R`, i.e, it first converts the type to bytes and that applies the `FnOnce` passed as an argument to produce the output `R`.

Reply:
But doesn't the default implementation of `Encode` allow us to use `encode()` when we implement `using_encoded`? 

TBH, I am not sure if this is important. Just was thinking that this is the case for how the default impl works.